### PR TITLE
Handle headers on loopback devices on bsd/osx

### PR DESCRIPTION
**Pull request by faern**
The previous version prepended 4 bytes == AF_INET to datalink sends. On my computer this header is added automatically so it doesn't seem to be needed. I didn't find any documentation describing this and I only tested it on my OS X 10.11.3 laptop, so I can't guarantee it works for every macos/freebsd flavor.

The previous version also kept the ethernet header. I don't think these are supposed to be included on the loopback interface. This PR changes so when sending to loopback on the datalink layer we discard the header and simply send the payload.

During read the OS will only give us the payload (prepended by the 4 bytes AF_INET) but the API force us to return an `EthernetPacket`. This PR changes so that the 14 bytes before the payload is zeroed out and included in the returned value. Thus, all packets read from a loopback device will have zeroes in all ethernet header fields.

@mrmonday

### CODE REVIEW

File: src/datalink/bpf.rs [hunk 1]
```diff
@@ -253,30 +261,43 @@ pub struct DataLinkChannelIteratorImpl<'a> {
...
                 }
             }
         }
-        let (start, len) = self.packets.pop_front().unwrap();
+        let (start, mut len) = self.packets.pop_front().unwrap();
+        len += buffer_offset;
+        // Zero out part that will become fake ethernet header if on loopback.
+        for i in (&mut self.pc.read_buffer[start..start + buffer_offset]).iter_mut() {
+            *i = 0;
+        }
```
Comment (faern) [inline src/datalink/bpf.rs]: Is there any shorter way to zero out a slice? It feels like there should be a nice functional style, one-liner laying around that I'm missing :)

File: src/datalink/bpf.rs [hunk 2]
```diff
@@ -27,11 +26,12 @@ use util::NetworkInterface;
 
 // NOTE buffer must be word aligned.
 #[inline]
-pub fn datalink_channel(network_interface: &NetworkInterface,
-                        write_buffer_size: usize,
-                        read_buffer_size: usize,
-                        channel_type: DataLinkChannelType)
-    -> io::Result<(Box<EthernetDataLinkSender>, Box<EthernetDataLinkReceiver>)> {
+pub fn datalink_channel
+    (network_interface: &NetworkInterface,
```
Comment (mrmonday) [inline src/datalink/bpf.rs]: Why this change? I don't like the arguments starting on the next line. I assume this was rustfmt's work?

File: src/datalink/bpf.rs (hunk 1 above)
Comment (mrmonday) [inline src/datalink/bpf.rs]: I agree, but I don't know of any way, off the top of my head.

**Comment by mrmonday** [comment]
Just to repeat what I said on IRC:

I don't like this, since it's fixing a hack which shouldn't be there in the first place.

However, you make a convincing point that it currently doesn't work at all, and this makes it sort-of work (much like the hack was intended to).

Other than the stylistic change, I'm happy (perhaps the wrong word :D) to pull it, with the hope that we can do it properly at a later point in time.

File: src/datalink/bpf.rs (hunk 2 above)
Comment (faern) [inline src/datalink/bpf.rs]: Yes it is. I obey rustfmt all the time. Actually my editor automatically runs it on every save, so this just ended up here. I could back it out but before doing that I would like to ask you to consider going full rustfmt on the entire project instead. It's so nice to remove the human aspect of code style and make it fully automatic. I don't love every line rustfmt gives me, but at least it will be consistent and I don't have to argue with anyone about it because I didn't make the decision :)

**Comment by faern** [comment]
Do you have any suggestion on how this proper solution would look? I think this is pretty OK right now since all information we have access to is the payload, there are no headers but still a datalink packet. I'm not sure what this protocol is named, should there be a different datalink packet type for this?

File: src/datalink/bpf.rs (hunk 2 above)
Comment (mrmonday) [inline src/datalink/bpf.rs]: I'm sure there's an option to stop it doing this in there - could you investigate? There's already a rustfmt configuration file in the repo.

I would love to start using rustfmt for everything, but unfortunately it's not currently possible (correct me if I'm wrong/this has changed) to exclude areas. Specifically, see https://github.com/libpnet/libpnet/blob/f08e08e76e6dc7498171982f58caa7b2c1270b89/src/packet/ipv4.rs.in#L263-L272 - last time I checked, rustfmt absolutely destroyed the formatting on that array literal, making it completely illegible.

If it doesn't do that any more, I'd be happy to start using it, and perhaps integrating it with the CI.

**Comment by mrmonday** [comment]
The proper solution is to fix issue #1, which I've been gradually chipping away at - there would be an alternative packet type for it (NullPacket, I imagine) which would be used instead.

File: src/datalink/bpf.rs (hunk 2 above)
Comment (faern) [inline src/datalink/bpf.rs]: Yes, it still destroys that special format. I'm divided here. Configuring the formatter brings back the human factor :( It would be so beautiful to remove that completely.
It's possible to add `#[cfg_attr(rustfmt, rustfmt_skip)]` to make it skip formating, but I only get that to work for entire files, not individual lines or functions.

I'll simply back out the reformatting.

**Comment by mrmonday** [comment]
@homu r+
